### PR TITLE
Possibility to dump tree as json

### PR DESCRIPTION
**Pull request by mieczkowski**
Hello,

I need possibility to dump tree to list of records (to store in DB for example), so I've added two methods:

- `MarshalJSON` to marshal tree (separated ipv4 and ipv6 keys) to json.

- `DumpList` to build custom json struct, use for example in database storage, or for debug purpose.

Test data is taken from `stringify_test.go`.

### CODE REVIEW

**Comment by gaissmai** [comment]
@mieczkowski thanks for contribution. Since you didn't open an issue before, I updated the main branch already by accident. Please rebase your PR on it.
I'm not happy with the public DumpList, please make it private and please check again the docstrings, I've seen at least one typo.
Best regards
  Charly

... and by reading the code (well done!) I've just some minor remarks:
- rename dumpList to dumpListRec
- rename DumpList to dumpList
- please change the signatures without error results, there is no error generating code in these fuctions

just my five cents

**Comment by mieczkowski** [comment]
Good catch with errors handling - I've simplified code. However I can not test one error check, because there is no chance to simulate this error, except some I/O errors inside `json.Marshal`:

```go
buf, err := json.Marshal(result)
if err != nil {
	return nil, err
}
```

Please consider leaving DumpList as public function. In my case I have list of subnets in DB (flat list). I need to rebuild tree after every operation on this with some validations, and:

1. Marshaling into json is helpful for rest api (returning sorted tree to frontend)
2. DumpList is helpful for add/remove operations and update DB

Without this method I need to marshal json and parse it into map[string]<smth> . Unnecessary waste of resources and without it I will end up with fork - and I don't want that ;)

File: jsonify.go [hunk 1]
```diff
@@ -44,40 +37,28 @@ func (t *Table[V]) MarshalJSON() ([]byte, error) {
 }
 
 // DumpList dumps ipv4 od ipv6 tree into list of roots and their children
```
Comment (gaissmai) [inline jsonify.go]: still a typo in this public docstring and a missing dot at end of line. If this method has to be public, please invest more time to find a clear description or spend an extra test or maybe a dumplist_example_test.

**Review by gaissmai — COMMENTED** [review]
please see my inline comment to DumpList

File: jsonify.go (hunk 1 above)
Comment (mieczkowski) [reply jsonify.go]: I've fixed both docstrings.

I don't think that we need separated tests for DumpList. It will be duplicate of json tests.

Comment (gaissmai) [reply jsonify.go]: The test for DumpList was not meant for testing, just to better explain the public method, but I can live without or add it myself later.

Two more small remarks:
- public type ListElement needs a docstring.
- please rename field Cidr to CIDR, for golang naming convention, like HTTP and not Http
Regards
   Charly

Comment (mieczkowski) [reply jsonify.go]: Done

**Comment by gaissmai** [comment]
What do you think about the following naming:

```go
type DumpListElement[V any] struct {}
func (t *Table[V]) DumpList(is4 bool) []DumpListElement[V]
```

background: the type only deals with dumping and marshaling, nothing with the real bart-Table algo.
or maybe: `DumpListNode`, since it's a recursive datastructure

**Comment by mieczkowski** [comment]
> What do you think about the following naming:
> 
> ```go
> type DumpListElement[V any] struct {}
> func (t *Table[V]) DumpList(is4 bool) []DumpListElement[V]
> ```
> 
> background: the type only deals with dumping and marshaling, nothing with the real bart-Table algo. or maybe: `DumpListNode`, since it's a recursive datastructure

I vote for `DumpListNode` , looks better than `element` since it's a tree :)

**Comment by gaissmai** [comment]
merged, thanks for contribution

**Pull request merged**